### PR TITLE
fix invoice render multiple tax rates

### LINKED ISSUE #12067: [Invoices] Several tax rates are showing up on the invoice for the same product
## Description
For some products, several tax rates are displayed on the invoice while only one is applied / can be applied.

Maybe linked / same error as in https://github.com/openfoodfoundation/openfoodnetwork/issues/12066

## Expected Behavior

We need to show only the tax rate that is used for this product

## Actual Behaviour
All tax rates are showing up

## Steps to Reproduce
Not sure yet how to end up in this situation, but on UK staging with invoices toggle ON, you can reproduce on this order for example: 

https://staging.openfoodnetwork.org.uk/admin/orders/R858337342/invoices

https://staging.openfoodnetwork.org.uk/admin/orders/R858337342/print?invoice_id=53

## Animated Gif/Screenshot

image

## Severity

bug-s3: a feature is broken but there is a workaround

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- Closes #12067

#### What should we test?

Here are the steps to reproduce the bug:

1. create a tax category (TG)
2. create multiple tax rates that belong to TG: TR1, TR2, TR3
3. create an order in a way that only one of the tax rates should apply to the line items (for example TR1) (TR2 & TR3 should not apply)
4. notice that all the tax rates will be listed in the invoice.
#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/invoice/data_presenter.rb [hunk 1]
```diff
@@ -91,6 +91,14 @@ def display_shipment_amount_with_taxes
       Spree::Money.new(shipment.amount + shipment.additional_tax_total, currency:)
     end
 
+    def display_line_item_tax_rate(item)
+      all_tax_adjustments.select { |a|
+        a.adjustable.type == 'Spree::LineItem' && a.adjustable.id == item.id
+      }.map(&:originator).map { |tr|
```
Comment (rioug) [inline app/models/invoice/data_presenter.rb:97]: Couldn't you use a query here ? something like :
```ruby
all_tax_adjustments..where(adjustable_type: "Spree::LineItem").where(adjustable_id: item.id)
  .map(&:originator) 
  #  ...
```

File: app/models/invoice/data_presenter.rb [hunk 2]
```diff
@@ -91,6 +91,14 @@ def display_shipment_amount_with_taxes
       Spree::Money.new(shipment.amount + shipment.additional_tax_total, currency:)
     end
 
+    def display_line_item_tax_rate(item)
```
Comment (rioug) [inline app/models/invoice/data_presenter.rb:94]: It would be good to have a test covering the scenario we are trying to fix

**Review by rioug — CHANGES_REQUESTED** [review]
Welcome back Mohamed ! Looks good, but I think we should add some testing. Moving back to Dev

File: app/models/invoice/data_presenter.rb (hunk 1 above)
Comment (abdellani) [reply app/models/invoice/data_presenter.rb:97]: I don't think that's possible.
This code runs inside the data presenter. `all_tax_adjustments` represents an array inside the serialized invoice.

File: app/models/invoice/data_presenter.rb (hunk 2 above)
Comment (abdellani) [reply app/models/invoice/data_presenter.rb:94]: I added new tests :)

File: app/models/spree/line_item.rb [hunk 3]
```diff
@@ -193,6 +193,7 @@ def added_tax
       adjustments.tax.additional.sum(:amount)
     end
 
+    # Some of the tax rates may not be applicable depending to the order's tax zone
```
Comment (abdellani) [inline app/models/spree/line_item.rb:196]: I think that the method's name is misleading.
It returns all the tax rates regardless if they are application or not.

**Comment by abdellani** [comment]
> Welcome back Mohamed

Thank you @rioug , happy to return back :partying_face:  
 
I updated my PR : )

File: app/models/invoice/data_presenter.rb (hunk 1 above)
Comment (rioug) [reply app/models/invoice/data_presenter.rb:97]: I see, I always forget the data presenter is working on serialised data.

**Review by rioug — APPROVED** [review]
Looks good  :+1:

File: app/models/spree/line_item.rb (hunk 3 above)
Comment (dacook) [reply app/models/spree/line_item.rb:196]: Good point. I just did a quick search and found at least two usages of this method, and I wonder if this bug would need to be fixed there also?

- [checkout_helper.rb](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/helpers/checkout_helper.rb#L73-L75)
- [sales_tax_totals_by_producer.rb](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/sales_tax/sales_tax_totals_by_producer.rb#L25-L33)

Are you able to replicate the issue in checkout and the sales tax report? If so, we should raise an issue for that too, or update this one.

**Review by dacook — APPROVED** [review]
Welcome back, and great job on replicating and squashing the bug! Looks like quite a specific case to set up and test.

I can see this fixes the raised issue in the invoice, so good to go. 
But I have a query if we need to go further..?

**Comment by RachL** [comment]
YAY this is fixed! Good one @abdellani :muscle:

I will ask Theresa to do a final test in prod. Merging!

image

**Pull request merged**